File: 백준/Gold/15684. 사다리 조작/사다리 조작.py
import sys
input = sys.stdin.readline

n, m, h = map(int, input().split())

graph = [[0] * n for _ in range(h)]

# 사다리왼쪽1, 오른쪽-1
for _ in range(m):
    a, b = map(int, input().split())
    graph[a-1][b-1] = 1
    graph[a-1][b] = -1

# i번째 사다리에서 시작해서 i번째 사다리에 도착하는지?
def check():
    global n, h
    for i in range(n-1):
        x, y = 0, i
        while x <= h-1:
            if graph[x][y] == 1:
                x+=1
                y+=1
            
            elif graph[x][y] == -1:
                x+=1
                y-=1

            else:
                x+=1

        if y != i:
            return False
    return True


ans = sys.maxsize
def radder(depth,x,y):
    global n, ans

    # 길이 3 넘으면 return
    if depth > 3:
        return
    # depth가 현재 답 이상일 경우 return(가지치기)
    if ans <= depth:
        return
    # check시 답 갱신 / 답이 1이면 바로 종료(가지치기)
    if check():
        ans = min(ans, depth)
        if ans == 1:
            print(1)
            exit()
        return

    # 사다리를 두는데 모든 경우의수를 체크한다. 하지만 위쪽 사다리를 점검 후 아래쪽 사다리를 점검중인데 굳이 위쪽에 사다리를 둘 필요가 없다. 중복되기 때문
    for i in range(x,h):
        # 같은 줄의 사다리를 검사할 때 앞쪽을 중복해서 체크할 필요는 없다.
        k = y if i == x else 0
        for j in range(k, n-1):
            if graph[i][j] == 0 and graph[i][j+1] == 0:                
                graph[i][j] = 1
                graph[i][j+1] = -1
                radder(depth+1,i,j+2)
                graph[i][j] = 0
                graph[i][j+1] = 0
                
if m == 0:
    print(0)
else:
    radder(0,0,0)
    if ans > 3:
        print(-1)
    else:
        print(ans)

File: 백준/Gold/15684. 사다리 조작/사다리 조작.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int n;
    static int m;
    static int h;
    static int[][] graph;
    static int ans = (int) 1e9;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken()); // 설치할 수 있는 사다리의 개수
        h = Integer.parseInt(st.nextToken());

        graph = new int[h][n];

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            graph[a - 1][b - 1] = -1; // 사다리 왼쪽은 -1
            graph[a - 1][b] = 1; // 사다리 오른쪽은 1
        }

        dfs(0, 0, 0);

        if (ans > 3) {
            ans = -1;
        }

        sb.append(ans);
        System.out.println(sb);
    }

    static void dfs(int radderCnt, int x, int y) {
        if (radderCnt > 3) { // 사다리 개수가 3개 넘으면 안된다.
            return;
        }

        if (radderCnt >= ans) { // 현재 답보다 더 많은 개수 설치하려고 하면 안된다.
            return;
        }

        if (check()) { // 검사 시 true라면 ans 갱신
            ans = Math.min(ans, radderCnt);
        }

        for (int i = x; i < h; i++) {
            for (int j = y; j < n - 1; j++) {
                if (graph[i][j] == 0 && graph[i][j + 1] == 0) {
                    graph[i][j] = -1;
                    graph[i][j + 1] = 1;

                    //사다리 설치 후 2칸 뒤부터 탐색
                    dfs(radderCnt + 1, i, j + 2);

                    graph[i][j] = 0;
                    graph[i][j + 1] = 0;
                }
            }
            y = 0; // 다음 행으로 넘어갈 때 0열 부터 탐색
        }

    }

    static boolean check() { // 원래 자기 사다리로 골인하는지??
        for (int i = 0; i < n; i++) {
            int x = 0;
            int y = i;
            while (x != h) {
                if (graph[x][y] == -1) {
                    y++;
                } else if (graph[x][y] == 1) {
                    y--;
                }
                x++;
            }
            if (i != y) {
                return false;
            }
        }
        return true;
    }
}